kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h

```
@@ -259,6 +259,7 @@ class ResidualBasedNewtonRaphsonStrategy
      
     virtual ~ResidualBasedNewtonRaphsonStrategy()
     {
+        Clear();
```

What is being cleaned here? If it is just mpA, mpDx, mpb, I don't know if this is necessary, since those are smart pointer and will self-destruct anyway if no-one references them.

---

Hey Jordi, a problem occurs with the MultiLevelSolver when the order of destruction is not carefully controlled. The member function Solve() creates a multilevel preconditioner of the matrix in mpA. This preconditioner stores a reference to the matrix which it tries to access when it is destructed. We can't prevent the matrix in mpA from destructing early by storing a smart pointer in the linear solver because the matrix is passed as a reference to Solve(). Therefore, we need to call the Clear() function of the linear solver before the matrix is freed.

---

Hello,

i have two comments on this
1 - the behaviour of the trilinos app should be as close as possible to the one of the serial one. 
2 - i think that calling Clear on destruction makes sense... why do you differ?

---

@msandre OK, the issue you mention is a good reason to force the clear. However, I think that having to call it here is indicative of confused ownership of stuff. I think it would make sense to store a copy of the mpA pointer in the solver/preconditioner, but I guess that if it was possible to do so without changing the interface you would have done so.

I'm not holding back the review for this (since it is beyond the scope of your fix), but we should keep it in mind going forward.

@RiccardoRossi  My reasoning for not calling Clear is that (ideally) it should do nothing: everything managed by this class should be correctly deallocated by the default destructor. Obviously this is not the case here, as @msandre noted.

---

@jcotela I agree that this is an exception to the standard of letting smart pointers do everything automatically. It is however necessary in this case if we want to be able to reuse the ML preconditioner without changing the interface. @RiccardoRossi if by point 1 you mean that the amgcl solver should behave similar to the ML then I don't think this can be fully achieved since the two solvers have different interfaces:

The ML preconditioner is built separately from the solver as

MLPreconditionerPointerType pPrec(new ML_Epetra::MultiLevelPreconditioner(rA, mMLParameterList, true));
Epetra_LinearProblem AztecProblem(&rA,&rX,&rB);
AztecOO aztec_solver(AztecProblem);
aztec_solver.SetPrecOperator(&(*pPrec));

The multigrid structure is built directly in the constructor of the amgcl solver 

auto p_built_in_matrix = amgcl::adapter::zero_copy(A.size1(), A.index1_data().begin(),
                A.index2_data().begin(), A.value_data().begin());

p_solver = boost::make_shared<solver_type>(
                amgcl::adapter::block_matrix<TBlockSize, block_type>(*p_builtin_matrix), prm); 

As I understand, the amgcl interface could be modified to allow storing the solver and reusing it with multiple right hand sides. If this is desired, I think it should be opened as a new issue.

---

@msandre I still can't see the problem with the ML preconditioner. I have tried to take a look into the trilinos ML code and they are apparently careful enough not to deallocate the rA. I would like to understand the issue better, because I suspect that this fix is hiding an error somewhere else. Do you have an example that failed without your fix?

---

@jcotela this is the test case I used.

https://gigamove.rz.rwth-aachen.de/d/id/es4QsU52KUvxEX?11&id=es4QsU52KUvxEX

Here is the output before the commit:

#0  0x00007fffe2fbde43 in ML_Ifpack_Destroy ()
   from /home/mike/Software/kratos/trilinos/lib/libml.so.11
#1  0x00007fffe2ed371f in ML_Smoother_Clean ()
   from /home/mike/Software/kratos/trilinos/lib/libml.so.11
#2  0x00007fffe2e7df8b in ML_Destroy ()
   from /home/mike/Software/kratos/trilinos/lib/libml.so.11
#3  0x00007fffe2fd8041 in ML_Epetra::MultiLevelPreconditioner::DestroyPreconditioner() () from /home/mike/Software/kratos/trilinos/lib/libml.so.11
#4  0x00007fffe2ff5075 in ML_Epetra::MultiLevelPreconditioner::~MultiLevelPreconditioner() () from /home/mike/Software/kratos/trilinos/lib/libml.so.11
#5  0x00007ffff40d7dfe in boost::detail::sp_counted_base::release() ()
   from /home/mike/Software/kratos/kratos/libs/libboost_python.so.1.59.0
#6  0x00007fffe4e42822 in ~shared_count (this=<optimized out>, 
    __in_chrg=<optimized out>)
    at /home/mike/Software/kratos/boost/include/boost/smart_ptr/detail/shared_count.hpp:467
#7  ~shared_ptr (this=<optimized out>, __in_chrg=<optimized out>)
    at /home/mike/Software/kratos/boost/include/boost/smart_ptr/shared_ptr.hpp:330
#8  reset (this=0x1545018)
    at /home/mike/Software/kratos/boost/include/boost/smart_ptr/shared_ptr.hpp:653
#9  ResetPreconditioner (this=0x1544eb0)

---

Thanks! I'll take a look